README.md
# Readme <a href='https://osf.io/zcvbs/'><img src='worcs_badge.png' align="right" height="139" /></a>

<!-- Please add a brief introduction to explain what the project is about    -->

## Where do I start?

You can load this project in Rstudio by opening the file called 

## Project structure

<!--  You can add rows to this table, using "|" to separate columns.         -->
File                        | Description                      | Usage         
--------------------------- | -------------------------------- | --------------
README.md                   | Description of project           | Human editable
manyanalysts_religion.Rproj | Project file                     | Loads project 
LICENSE                     | User permissions                 | Read only     
.worcs                      | WORCS metadata YAML              | Read only     
prepare_data.R              | Script to process raw data       | Human editable
manuscript/manuscript.rmd   | Source code for paper            | Human editable
manuscript/references.bib   | BibTex references for manuscript | Human editable
renv.lock                   | Reproducible R environment       | Read only     

<!--  You can consider adding the following to this file:                    -->
<!--  * A citation reference for your project                                -->
<!--  * Contact information for questions/comments                           -->
<!--  * How people can offer to contribute to the project                    -->
<!--  * A contributor code of conduct, https://www.contributor-covenant.org/ -->

# Reproducibility

This project uses the Workflow for Open Reproducible Code in Science (WORCS) to
ensure transparency and reproducibility. The workflow is designed to meet the
principles of Open Science throughout a research project. 

* To learn how WORCS helps researchers meet the TOP-guidelines and FAIR principles, read the preprint at https://osf.io/zcvbs/
* To get started with `worcs`, see the [setup vignette](https://cjvanlissa.github.io/worcs/articles/setup.html)
* For detailed information about the steps of the WORCS workflow, see the [workflow vignette](https://cjvanlissa.github.io/worcs/articles/workflow.html)
* For a brief overview of the steps of the WORCS workflow, see below.

## WORCS: Steps to follow for a project

## Phase 1: Study design

1. Create a new (Public or Private) repository on 'GitHub'
2. Create a new RStudio project using the WORCS template
3. Optional: Preregister your analysis
4. Optional: Upload preregistration to another repository
5. Optional: Add study Materials to the repository

## Phase 2: Data analysis

6. Load the raw data
7. Save the data using `open_data()` or `closed_data()`. Never commit data to 'Git' that you do not intend to share
8. Write the manuscript in `manuscript.Rmd`, using code chunks to perform the analyses.
9. Commit every small change
10. Cite essential references with `@`, and non-essential references with `@@`

## Phase 3: Submission/publication

11. Store the R environment by calling `renv::snapshot()`
12. Optional: Add a WORCS-badge to your README file and complete the optional elements of the WORCS checklist
13. Make the Private 'GitHub' repository Public
14. [Create a project page on the Open Science Framework](https://help.osf.io/hc/en-us/articles/360019737594-Create-a-Project)
15. Connect your 'OSF' project page to the 'GitHub' repository](https://help.osf.io/hc/en-us/articles/360019929813-Connect-GitHub-to-a-Project)
16. Add an open science statement to the Abstract or Author notes, which links to the 'GitHub' repository or 'OSF' page
17. Knit the paper to PDF
18. Optional: Publish a preprint
19. Submit the paper, and tag the release of the submitted paper as in Step 3.

## Notes for cautious researchers

Some researchers might want to share their work only once the paper is accepted for publication. In this case, we recommend creating a "Private" repository in Step 1, and completing Steps 13-18 upon acceptance.


## Access to data

Some of the data used in this project are not publically available.
Synthetic data with similar characteristics to the original data have been provided. Using the function load_data() will load these synthetic data when the original data are unavailable. Note that these synthetic data cannot be used to reproduce the original results. However, it does allow users to run the code and, optionally, generate valid code that can be evaluated using the original data by the project authors.
To request access to the original data, [open a GitHub issue](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/creating-an-issue).

<!--Clarify here how users should contact you to gain access to the data, or to submit syntax for evaluation on the original data.-->


## Access to data

Some of the data used in this project are not publically available.
To request access to the original data, [open a GitHub issue](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/creating-an-issue).

<!--Clarify here how users should contact you to gain access to the data, or to submit syntax for evaluation on the original data.-->


## Access to data

Some of the data used in this project are not publically available.
To request access to the original data, [open a GitHub issue](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/creating-an-issue).

<!--Clarify here how users should contact you to gain access to the data, or to submit syntax for evaluation on the original data.-->


changes.txt
* Misinterpreted comment about missing values in data documentation
* Accidentally put both the original variable ethicity and the derived variable "majority" in the analysis; removed these
* Tried using ESTIMATOR = BAYES for main models because integration did not converge
* Removed categorical items, because no DIC is available for cat outcomes http://www.statmodel.com/discussion/messages/9/6256.html?1600907976
* DIC is not available for latent variable interactions, thus it is not possible to determine best fitting model according to preregistration
* Decided to conduct path analysis with mean score scales of the observed variables to determine best fitting model, then revert back to Bayes for estimation of the final model
* Final model does not converge with Bayes; decided to use path analysis with mean scores for all analyses

analysis.R
library(worcs)
library(missRanger)
library(MplusAutomation)
library(tidySEM)
library(tuneRanger)
library(ranger)
library(metaforest)
# Load data
load_data()

# Set aside testing sample for unbiased evaluation of model fit
set.seed(893)
train <- sample(1:nrow(df), size = floor(.70*nrow(df)))
df_test <- df[-train, ]
df <- df[train, ]

# Remove failed attention check
df <- df[df$attention_check == 1, ]

# Impute missings
df <- missRanger(df)

# The models below are subject to changes - only when using the training data
syn_m1 <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "well WITH rel@0;",
  "well WITH cnorm@0;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "[sint];",
  "sreli;",
  "snorm;",
  "sint;",
  )

m1 <- mplusObject(TITLE = "M1",
                   VARIABLE = c(
                     "CATEGORICAL = rel_3 rel_4;",
                     "WITHIN = age gender ses education ethnicity majority",
                     "rel_1-rel_9",
                     "cnorm_1-cnorm_2",
                     "wb_phys_1-wb_phys_7",
                     "wb_psych_1-wb_psych_6",
                     "wb_soc_1-wb_soc_3;",
                     "CLUSTER IS clus;"
                   ),
                   ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                                "ALGORITHM = INTEGRATION;",
                                "INTEGRATION = 10;"),
                   MODEL = syn_m1,
                   rdata = df)

syn_m1fx <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "well WITH rel@0;",
  "well WITH cnorm@0;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "[sint];",
  "sreli@0;",
  "snorm@0;",
  "sint@0;",
)

m1fx <- mplusObject(TITLE = "M1fx",
                  VARIABLE = c(
                    "CATEGORICAL = rel_3 rel_4;",
                    "WITHIN = age gender ses education ethnicity majority",
                    "rel_1-rel_9",
                    "cnorm_1-cnorm_2",
                    "wb_phys_1-wb_phys_7",
                    "wb_psych_1-wb_psych_6",
                    "wb_soc_1-wb_soc_3;",
                    "CLUSTER IS clus;"
                  ),
                  ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                               "ALGORITHM = INTEGRATION;",
                               "INTEGRATION = 10;"),
                  MODEL = syn_m1fx,
                  rdata = df)

syn_m2 <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "sreli | well ON rel;",
  "snorm | well ON cnorm;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "sreli;",
  "snorm;"
)

m2 <- mplusObject(TITLE = "M2",
                  VARIABLE = c(
                    "CATEGORICAL = rel_3 rel_4;",
                    "WITHIN = age gender ses education ethnicity majority",
                    "rel_1-rel_9",
                    "cnorm_1-cnorm_2",
                    "wb_phys_1-wb_phys_7",
                    "wb_psych_1-wb_psych_6",
                    "wb_soc_1-wb_soc_3;",
                    "CLUSTER IS clus;"
                  ),
                  ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                               "ALGORITHM = INTEGRATION;",
                               "INTEGRATION = 10;"),
                  MODEL = syn_m2,
                  rdata = df)

syn_m2fx <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "sreli | well ON rel;",
  "snorm | well ON cnorm;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "sreli@0;",
  "snorm@0;"
)

m2fx <- mplusObject(TITLE = "M2fx",
                  VARIABLE = c(
                    "CATEGORICAL = rel_3 rel_4;",
                    "WITHIN = age gender ses education ethnicity majority",
                    "rel_1-rel_9",
                    "cnorm_1-cnorm_2",
                    "wb_phys_1-wb_phys_7",
                    "wb_psych_1-wb_psych_6",
                    "wb_soc_1-wb_soc_3;",
                    "CLUSTER IS clus;"
                  ),
                  ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                               "ALGORITHM = INTEGRATION;",
                               "INTEGRATION = 10;"),
                  MODEL = syn_m2fx,
                  rdata = df)


syn_m3 <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "int | rel XWITH cnorm;",
  "sreli | well ON rel;",
  "snorm | well ON cnorm;",
  "sint | well ON int;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "[sint];",
  "sreli;",
  "snorm;",
  "sint;",
)

m3 <- mplusObject(TITLE = "M3",
                  VARIABLE = c(
                    "CATEGORICAL = rel_3 rel_4;",
                    "WITHIN = age gender ses education ethnicity majority",
                    "rel_1-rel_9",
                    "cnorm_1-cnorm_2",
                    "wb_phys_1-wb_phys_7",
                    "wb_psych_1-wb_psych_6",
                    "wb_soc_1-wb_soc_3;",
                    "CLUSTER IS clus;"
                  ),
                  ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                               "ALGORITHM = INTEGRATION;",
                               "INTEGRATION = 10;"),
                  MODEL = syn_m3,
                  rdata = df)

syn_m3fx <- c(
  "%WITHIN%",
  "rel BY rel_1-rel_9;",
  "cnorm BY cnorm_1-cnorm_2;",
  "well BY wb_phys_1-wb_phys_7;",
  "well BY wb_psych_1-wb_psych_6;",
  "well BY wb_soc_1-wb_soc_3;",
  "int | rel XWITH cnorm;",
  "sreli | well ON rel;",
  "snorm | well ON cnorm;",
  "sint | well ON int;",
  "well ON age;",
  "well ON gender;",
  "well ON ses;",
  "well ON education;",
  "well ON ethnicity;",
  "well ON majority;",
  
  "%BETWEEN%",
  "[sreli];",
  "[snorm];",
  "[sint];",
  "sreli@0;",
  "snorm@0;",
  "sint@0;",
)

m3fx <- mplusObject(TITLE = "M3fx",
                  VARIABLE = c(
                    "CATEGORICAL = rel_3 rel_4;",
                    "WITHIN = age gender ses education ethnicity majority",
                    "rel_1-rel_9",
                    "cnorm_1-cnorm_2",
                    "wb_phys_1-wb_phys_7",
                    "wb_psych_1-wb_psych_6",
                    "wb_soc_1-wb_soc_3;",
                    "CLUSTER IS clus;"
                  ),
                  ANALYSIS = c("TYPE = TWOLEVEL RANDOM;",
                               "ALGORITHM = INTEGRATION;",
                               "INTEGRATION = 10;"),
                  MODEL = syn_m3fx,
                  rdata = df)


# Once models are finalized, evaluate the set on only the testing data.
# Then, compute fit indices and likelihood ratio tests on the results of the
# testing data.
MplusAutomation::SummaryTable()
tidySEM:::SB_chisq_Pvalues()

# Select the definitive model, then evaluate that model on the full dataset
# and perform inference.

# Random forests ----------------------------------------------------------

df_rf <- df
names(df_rf) <- gsub("^([a-z]+)_([a-z]+)$", "\\1\\2", names(df_rf))
dict <- tidySEM::tidy_sem(df_rf)
dict$dictionary$scale <- gsub("^(wb)_.*$", "wb", dict$dictionary$scale)

scales <- create_scales(dict)
df_rf <- cbind(df_rf, scales$scores)
names(df_rf)
df_rf[grepl("((rel|wb|cnorm)_|gdp|attention|subject)", names(df_rf))] <- NULL
df_rf <- df_rf[c("wb", names(df_rf)[!names(df_rf) == "wb"])]
set.seed(6983)
conv <- ranger(as.formula(paste0("wb ~ ", paste0(names(df_rf)[!names(df_rf) == "wb"], collapse = " + "))),  df_rf, num.trees = 5000)
plot(conv, data = df_rf)
task <- makeRegrTask(data = df_rf, target = "wb")

# Tuning
tune <- tuneRanger(task, num.trees = 100, num.threads = 2, iters = 70, save.file.path = NULL, parameters = list(replace = FALSE, respect.unordered.factors = "order", importance = "permutation"))

final <- tune$model$learner.model
metaforest::VarImpPlot(final)
PartialDependence(final, data = df_rf)

PartialDependence(final, vars = "rel", moderator = "cnorm", data = df_rf)

mplus_scales.R
# The models below are subject to changes - only when using the training data
library(tidySEM)
library(MplusAutomation)
run_scale_models <- TRUE
df_score <- df
dict <- tidy_sem(df)
dict$dictionary$scale[c(5:6, 32:47)] <- NA
dict$dictionary$scale <- gsub("^wb_(gen|phys|psych|soc)$", "well", dict$dictionary$scale)
scalescores <- tidySEM::create_scales(dict)
scalescores <- scale(scalescores$scores, scale = FALSE)
df_score <- cbind(df_score, scalescores)
df_score$int <- df_score$rel * df_score$cnorm

if(run_scale_models | !file.exists("m1_scale.out")){
  syn_m1_scale <- c(
    "%WITHIN%",
    "well ON age;",
    "well ON gender;",
    "well ON ses;",
    "well ON education;",
    "well ON majority;",
    "%BETWEEN%",
    "[well];"
  )
  #298705.124
  m1_scale <- mplusObject(TITLE = "M1_scale",
                          VARIABLE = paste0(c(
                            "WITHIN = age gender ses education majority",
                            ";",
                            "CLUSTER IS country;"
                          ), collapse = "\n"),
                          ANALYSIS = c("TYPE = TWOLEVEL RANDOM;"
                          ),
                          MODEL = syn_m1_scale,
                          #OUTPUT = "stdyx;",
                          rdata = df_score)
  res_m1_scale <- mplusModeler(m1_scale, modelout = "m1_scale.inp", run = 1L)$results
} else {
  res_m1_scale <- readModels("m1_scale.out")
}

if(run_scale_models | !file.exists("m2_scale.out")){
  syn_m2_scale <- c(
    "%WITHIN%",
    "sreli | well ON rel;",
    "snorm | well ON cnorm;",
    "well ON age;",
    "well ON gender;",
    "well ON ses;",
    "well ON education;",
    "well ON majority;",
    
    "%BETWEEN%",
    "[well];",
    "[sreli];",
    "[snorm];",
    "sreli;",
    "snorm;"
  )
  
  m2_scale <- mplusObject(TITLE = "M2_scale",
                          VARIABLE = paste0(c(
                            "WITHIN = age gender ses education majority",
                            "rel",
                            "cnorm",
                            ";",
                            "CLUSTER IS country;"
                          ), collapse = "\n"),
                          ANALYSIS = c("TYPE = TWOLEVEL RANDOM;"),
                          MODEL = syn_m2_scale,
                          rdata = df_score)
  
  res_m2_scale <- mplusModeler(m2_scale, modelout = "m2_scale.inp", run = 1L)$results
} else {
  res_m2_scale <- readModels("m2_scale.out")
}

if(run_scale_models | !file.exists("m2fx_scale.out")){
  syn_m2fx_scale <- c(
    "%WITHIN%",
    "sreli | well ON rel;",
    "snorm | well ON cnorm;",
    "well ON age;",
    "well ON gender;",
    "well ON ses;",
    "well ON education;",
    "well ON majority;",
    
    "%BETWEEN%",
    "[well];",
    "[sreli];",
    "[snorm];",
    "sreli@0;",
    "snorm@0;"
  )
  
  m2fx_scale <- mplusObject(TITLE = "M2fx_scale",
                            VARIABLE = paste0(c(
                              "WITHIN = age gender ses education majority",
                              "rel",
                              "cnorm",
                              ";",
                              "CLUSTER IS country;"), collapse = "\n"
                            ),
                            ANALYSIS = c("TYPE = TWOLEVEL RANDOM;"),
                            MODEL = syn_m2fx_scale,
                            rdata = df_score)
  
  res_m2fx_scale <- mplusModeler(m2fx_scale, modelout = "m2fx_scale.inp", run = 1L)$results
} else {
  res_m2fx_scale <- readModels("m2fx_scale.out")
}

if(run_scale_models | !file.exists("m3_scale.out")){
  syn_m3_scale <- c(
    "%WITHIN%",
    "sreli | well ON rel;",
    "snorm | well ON cnorm;",
    "sint | well ON int;",
    "well ON age;",
    "well ON gender;",
    "well ON ses;",
    "well ON education;",
    "well ON majority;",
    
    "%BETWEEN%",
    "[well];",
    "[sreli];",
    "[snorm];",
    "[sint];",
    "sreli;",
    "snorm;",
    "sint;"
  )
  
  m3_scale <- mplusObject(TITLE = "M3_scale",
                          VARIABLE = paste0(c(
                            "WITHIN = age gender ses education majority",
                            "rel",
                            "cnorm",
                            "int;",
                            "CLUSTER IS country;"), collapse = "\n"
                          ),
                          ANALYSIS = c("TYPE = TWOLEVEL RANDOM;"),
                          MODEL = syn_m3_scale,
                          OUTPUT = "TECH3;",
                          rdata = df_score)
  
  res_m3_scale <- mplusModeler(m3_scale, modelout = "m3_scale.inp", run = 1L)$results
} else {
  res_m3_scale <- readModels("m3_scale.out")
}

if(run_scale_models | !file.exists("m3fx_scale.out")){
  syn_m3fx_scale <- c(
    "%WITHIN%",
    "sreli | well ON rel;",
    "snorm | well ON cnorm;",
    "sint | well ON int;",
    "well ON age;",
    "well ON gender;",
    "well ON ses;",
    "well ON education;",
    "well ON majority;",
    
    "%BETWEEN%",
    "[well];",
    "[sreli];",
    "[snorm];",
    "[sint];",
    "sreli@0;",
    "snorm@0;",
    "sint@0;"
  )
  
  m3fx_scale <- mplusObject(TITLE = "M3fx_scale",
                            VARIABLE = paste0(c(
                              "WITHIN = age gender ses education majority",
                              "rel",
                              "cnorm",
                              "int;",
                              "CLUSTER IS country;"), collapse = "\n"
                            ),
                            ANALYSIS = c("TYPE = TWOLEVEL RANDOM;"),
                            
                            MODEL = syn_m3fx_scale,
                            rdata = df_score)
  
  res_m3fx_scale <- mplusModeler(m3fx_scale, modelout = "m3fx_scale.inp", run = 1L)$results
} else {
  res_m3fx_scale <- readModels("m3fx_scale.out")
}

results <- mget(grep("^res_.+scale", ls(), value = TRUE))

tmp <- MplusAutomation::SummaryTable(results, sortBy = "Title", keepCols = c("Title", "Parameters", "LL", "LLCorrectionFactor", "AIC", "BIC"), type = "none")

tmp$Title <- gsub("(\\d)_scale", "\\1 free", tmp$Title)
tmp$Title <- gsub("fx_scale", " fixed", tmp$Title)
tmp <- tmp[order(tmp$Parameters), ]
library(ggplot2)
p <- ggplot(data.frame(data.frame(Model = ordered(tmp$Title, levels = tmp$Title),
                                  BIC = tmp$BIC)),
            aes(x = Model, y = BIC)) + geom_path(group = 1) +
  geom_point()  + 
  geom_vline(xintercept = tmp$Title[which.min(tmp$BIC)], linetype = 2) +
  geom_hline(yintercept = tmp$BIC[which.min(tmp$BIC)], linetype = 2) +
  theme_bw()


m <- tmp[1:4, c("Parameters", "LL", "LLCorrectionFactor")]
mp1 <- tmp[2:5, c("Parameters", "LL", "LLCorrectionFactor")]

cd = ((m$Parameters*m$LLCorrectionFactor) - (mp1$Parameters*mp1$LLCorrectionFactor)) / (m$Parameters-mp1$Parameters)
degfree = mp1$Parameters - m$Parameters
TRd = -2*(m$LL - mp1$LL) / cd
tmp$LR_p <- c(NA, formatC(pchisq(TRd, degfree, lower.tail = FALSE), digits = 3, format = "f"))


prepare_data.R
# In this file, write the R-code necessary to load your original data file
# (e.g., an SPSS, Excel, or SAS-file), and convert it to a data.frame. Then,
# use the function open_data(your_data_frame) or closed_data(your_data_frame)
# to store the data.

library(worcs)

# Load data

# replace this with code to load data
df <- read.csv("MARP_data.csv", stringsAsFactors = FALSE)
# Recode missing data

# NOTE: Unlike indicated in the data documentation, none of the missing
# values described below are present in the data.
# Missing Values 
# 
# 
# The dataset contains missing values for the following variables: 
#   ● wb_soc_3 995 missing (did not want to disclose) 
# df$wb_soc_3[which(df$wb_soc_3 == 995)] <- NA
# # ● age 27 missing 
# df$age[which(df$age == 27)] <- NA
# # ● ses 5 missing 
# df$ses[which(df$ses == 5)] <- NA
# # ● ethnicity 409 missing (in France, this question was not asked due to legal 
# #                          restrictions; no missing values in other countries)  
# df$ethnicity[which(df$ethnicity == 409)] <- NA
# # ● denomination 5676 missing (this was only asked when participants indicated to 
# #                              belong to a religious denomination [rel_4]) 
# df$denomination[which(df$denomination == 5676)] <- NA
# # There are no missing data for the remaining variables.  


# Not religious and atheist same for present purposes
df$rel_3 <- df$rel_3 + 1
df$rel_3[which(df$rel_3 %in% c(1.5, 2))] <- 0
# Recode to dummy
df$rel_4[which(df$rel_4 %in% c(2))] <- 0

# Set gender == other to NA, because number is likely to low to use as additional moderator
df$gender[which(df$gender == "other")] <- NA

# For each country, determine the majority ethnicity
majority <- tapply(df$ethnicity, df$country, function(x){
  tab <- table(x)
  names(tab)[which.max(tab)]
})
# Create variable to use in analysis: Is the participant part of the majority ethnicity in their country?
df$majority <- as.numeric(df$ethnicity == majority[df$country])

# Check data for outliers (out of range as specified in data documentation)
desc <- worcs::descriptives(df)
write.csv(desc, "descriptives.csv", row.names = FALSE)

# Save data
imp_fun <- function(x){
  if(is.data.frame(x)){
    return(data.frame(sapply(x, imp_fun)))
  } else {
    out <- x
    if(inherits(x, "numeric")){
      out[is.na(out)] <- median(x[!is.na(out)])
    } else {
      out[is.na(out)] <- names(sort(table(out), decreasing = TRUE))[1]
    }
    out
  }
}
closed_data(df, model_expression = NULL,
            predict_expression = sample(y, size = length(y), replace = TRUE),
            missingness_expression = imp_fun(data))

fake_data.R
set.seed(3634)
n <- 10535
countries <- 24

df <- data.frame(subject = 1:n)

coun <- as.numeric(substring(as.character(abs(rnorm(n))*100),1,2))
keep <- unique(coun)[1:countries]
df$country <- findInterval(coun, sort(keep)[-24]) + 1
table(df$country)


# Religiosity 
# Personal religiosity was measured with 9 items. All items have been reverse-coded and 
# transformed onto a 0-1 scale, with 0 indicating low religiosity to 1 indicating high religiosity. 
# Values in round brackets () give the original response scale, [R] indicates that the item was 
# originally contra-indicative and has been reverse-coded.  
# 3. rel_1 frequency of service attendance (1-7) [R] 
# 4. rel_2 frequency of prayer (1-8) [R] 
# 5. rel_3 self-identification (1= religious, 2= not religious, 3=atheist) [R] 
# 6. rel_4 belong to denomination (1=yes, 2=no) [R] 
# 7. rel_5 belief in God/Gods (1-7) 
# 8. rel_6 belief in afterlife (1-7) 
# 9. rel_7 spirituality (1-7) 
# 10. rel_8 importance of religious lifestyle (1-5) 
# 11. rel_9 importance of belief in God/Gods (1-5)

df$rel_1 <- scale(sample.int(7, n, replace = TRUE))
df$rel_2 <- scale(sample.int(8, n, replace = TRUE))
df$rel_3 <- scale(sample.int(3, n, replace = TRUE))
df$rel_4 <- scale(sample.int(2, n, replace = TRUE))
df$rel_5 <- scale(sample.int(7, n, replace = TRUE))
df$rel_6 <- scale(sample.int(7, n, replace = TRUE))
df$rel_7 <- scale(sample.int(7, n, replace = TRUE))
df$rel_8 <- scale(sample.int(5, n, replace = TRUE))
df$rel_9 <- scale(sample.int(5, n, replace = TRUE))


# Cultural Norms of Religiosity  
# Perceived cultural norms of religiosity were measured with 2 items. Both items are transformed 
# onto a 0-1 scale, with 0 indicating low perceived cultural importance of religiosity and 1 
# indicating high perceived cultural importance of religiosity. Values in round brackets () give the original response scale. 
# 12. cnorm_1  importance of religious lifestyle for average person in country (1-5) 
# 13. cnorm_2 importance of belief in God/Gods for average person in country (1-5) 
df$cnorm_1 <- scale(sample.int(5, n, replace = TRUE))
df$cnorm_2 <- scale(sample.int(5, n, replace = TRUE))



# Well-Being  
# Well-being was measured with 18 items using a 1-5 response scale, with 1 indicating low 
# well-being and 5 indicating high well-being. The scale consists of 2 general well-being items, 7 
# physical well-being items, 6 psychological well-being items, and 3 social well-being items. [R] 
# indicates that the item was originally contra-indicative and has been reverse-coded.  
df$wb_gen_1 <- sample.int(5, n, replace = TRUE) # 14.  wb_gen_1 quality of life general 
df$wb_gen_2 <- sample.int(5, n, replace = TRUE) # 15.  wb_gen_2 satisfaction with health 


df$wb_phys_1 <- sample.int(5, n, replace = TRUE) # 16.  wb_phys_1 pain [R] 
df$wb_phys_2 <- sample.int(5, n, replace = TRUE) # 17.  wb_phys_2 medical dependence [R] 
df$wb_phys_3 <- sample.int(5, n, replace = TRUE) # 18.  wb_phys_3 energy 
df$wb_phys_4 <- sample.int(5, n, replace = TRUE) # 19.  wb_phys_4 mobility 
df$wb_phys_5 <- sample.int(5, n, replace = TRUE) # 20.  wb_phys_5 sleep 
df$wb_phys_6 <- sample.int(5, n, replace = TRUE) # 21.  wb_phys_6 activities 
df$wb_phys_7 <- sample.int(5, n, replace = TRUE) # 22.  wb_phys_7 work ability 
df$wb_psych_1 <- sample.int(5, n, replace = TRUE) # 23.  wb_psych_1 enjoying life 
df$wb_psych_2 <- sample.int(5, n, replace = TRUE) # 24.  wb_psych_2 meaningfulness 
df$wb_psych_3 <- sample.int(5, n, replace = TRUE) # 25.  wb_psych_3 concentration 
df$wb_psych_4 <- sample.int(5, n, replace = TRUE) # 26.  wb_psych_4 satisfaction physical appearance 
df$wb_psych_5 <- sample.int(5, n, replace = TRUE) # 27.  wb_psych_5 self-esteem 
df$wb_psych_6 <- sample.int(5, n, replace = TRUE) # 28.  wb_psych_6 negative affect [R] 
df$wb_soc_1 <- sample.int(5, n, replace = TRUE) # 29.  wb_soc_1 personal relations 
df$wb_soc_2 <- sample.int(5, n, replace = TRUE) # 30.  wb_soc_2 social support 
df$wb_soc_3 <- sample.int(5, n, replace = TRUE) # 31.  wb_soc_3 sexual satisfaction (note: NA indicates not disclosed) 

# 32. wb_overall_mean average of all 18 well-being items 
# 33. wb_phys_mean average of physical well-being subscale 
# 34. wb_psych_mean average of psychological well-being subscale 
# 35. wb_soc_mean average of social well-being subscale 
# 


#Demographics  
df$age <- round(runif(n, 18, 80)) #36. age age in numbers
df$gender <- sample.int(3, n, replace= TRUE, prob = c(.495, .495, .01)) #36. age age in numbers 
#37. gender (1=man, 2=woman, 3=other) 
df$ses <- sample.int(10, n, replace = TRUE) #38. ses subjective socioeconomic status as ladder (1 = bottom, 10 = top) 
df$education <- sample.int(7, n, replace = TRUE)#39. education education in levels (largely equivalent across samples; 1-7, 1 indicating the lowest level of education, 7 indicting the highest level of education) 
df$ethnicity <- sample.int(17, n, replace = TRUE) #40. ethnicity ethnicity (different per sample; 17 categories) 
#41. denomination if indicated member of religious denomination (different per sample; 20 categories) 
df$denomination <- sample.int(40, n, replace = TRUE)



#Additional Study Details 
df$gdp <- runif(countries, 10000, 100000)[df$country]#42. gdp gross domestic product per capita (in US$, data from 2017; country-level) 
df$gdp_scaled <- scale(df$gdp) # 43. gdp_scaled GDP standardized across the 24 countries (country-level) 


df$sample_type <- sample(c("gp", "students", "online", "mixed"), n, replace = TRUE) # 44. sample_type type of sample (general public, students, online panel, mixed) 
df$sample_type <- sample(c("credit", "money", "none", "raffle", "mixed"), n, replace = TRUE) #45. compensation participant reward (course credit, money, no compensation, raffle, mixed) 
df$attention_check <- sample.int(2, n, prob = c(.9, .1), replace = TRUE)#46. attention_check attention check (1=passed) 


rm(coun, countries, keep, n)

